[limshaoqi] iP

limshaoqi opened the pull request:
# ElChino - Un Gestor de Tareas Simple 📝
> "Friends can come and go, but banners hang forever." - Kobe Bryant [source](https://www.espn.com/blog/los-angeles/lakers/post/_/id/41673/kobe-friends-can-come-and-go-but-banners-hang-forever)

## Características
- Gestión de tareas basada en comandos
- Fácil de aprender
- Respetuoso del medio ambiente

## Todo lo que necessitas hacer es:
1. Descárgalo desde aquí.
2. Haz doble clic en él.
3. Añade tus tareas.
4. Deja que administre tus tareas por ti. 😉

### RELAK LAH
- [x] baik uh sial
- [ ] *relak*
- [ ] __rizz__ & ~~chill~~

```
"Deng bu dao tian hei, Yan huo bu hui tai wan mei.
Hui yi shao cheng hui, Hai shi deng bu dao jie wei.
Ta ceng shuo de wu suo wei, Wo pa yi tian yi tian bei cui hui - Rizzalations 24:7
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
renamed Duke to El Chino and added custom messages

Commit:
added echo user input and exit when 'bye'

Commit:
added list

Commit:
added Task.java

Commit:
added mark/umark feature

Commit:
made Task abstract & added Todo class

Commit:
added Deadline class

Commit:
added Event class

Commit:
added getTaskType method and used Stringformat method for Task class

Commit:
implemented getTaskType in Todo and Event classes from Task class

Commit:
refactored Elchino to work with TaskManager

Commit:
added TaskManager class to handle tasks

Commit:
added the test cases

Commit:
Add ElchinoException

Commit:
Add InvalidInputException

Commit:
Add InvalidCommandException

Commit:
Add EmptyDescriptionException

Commit:
add parseTaskNumber method and put exceptions in same package

Commit:
Updated printTaskList to print appropriate msg for empty list

Commit:
Add addTodo method

Commit:
Fixed addTodo input bug

Commit:
Updated addDeadline to handle exceptions

Commit:
Updated addEvent to handle exceptions

Commit:
Updated parseTaskNumber

Commit:
Added markTask method

Commit:
Added unmarkTask method

Commit:
Fixed TaskManager bug not handling empty input

Commit:
Updated addDeadline and addEvent to handle invalidinput exception case when not using /by or /from

Commit:
Add deleteTask method

Commit:
Add Storage.java to manage saving tasks

Commit:
Add save, load, parase methods

Commit:
Fix Storage.java bugs

Commit:
Merge branch 'branch-Level-7'

Commit:
Add date time parser

Commit:
Refactor Deadline and Event

Commit:
Merge branch 'branch-Level-8'

Commit:
Add Ui.java

Commit:
Add TaskList.java

Commit:
Add Parser.java

Commit:
Add Command.java

Commit:
Add ListCommand.java

Commit:
Add MarkCommand.java

Commit:
Add UnmarkCommand.java

Commit:
Add AddTodoCommand.java

Commit:
Add AddDeadlineCommand.java

Commit:
Add AddEventCommand.java

Commit:
Add DeleteCommand.java

Commit:
Add ExitCommand.java

Commit:
Add InvalidCommand.java

Commit:
Refactor Elchino.java

Commit:
Refactor commands messages

Commit:
Merge branch 'A-MoreOOP'

Commit:
Organize into Packages

Commit:
Update packages

Commit:
Merge tag 'A-Packages'

Commit:
Merge remote-tracking branch 'origin/add-gradle-support'

Commit:
Update gradle entry point

Commit:
Add JUnit tests for TaskList and Parser

Commit:
Update build.gradle shadowJar configuration

Commit:
Add JavaDocs for exceptions and commands

Commit:
Add JavaDocs for Elchino, tasks, storage and ui

Commit:
Reformat to follow coding standards

Commit:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Add FindCommand

Commit:
Merge branch 'branch-Level-9'

Frozennfishh commented on src/main/java/elchino/Elchino.java:
```diff
@@ -0,0 +1,86 @@
+package elchino;
+
+import java.util.ArrayList;
+
+import elchino.exceptions.*;
+
+import elchino.ui.Ui;
+
+import elchino.tasks.*;
+
+import elchino.storage.Storage;
+
+import elchino.parser.Parser;
+
+import elchino.commands.*;
```
Perhaps the import files of the same type could be group together to be able to more easily check?

limshaoqi replied:
vale

Frozennfishh commented on src/main/java/elchino/data/tasks.txt:
Maybe can remove redundant empty files like this

limshaoqi replied:
bet

Frozennfishh commented on src/main/java/elchino/tasks/Task.java:
```diff
@@ -0,0 +1,81 @@
+     * @return The LocalDateTime object representing the date.
+     * throws IllegalArgumentException If the date string is invalid.
+     */
+    protected static LocalDateTime parseDate(String date) {
+        try {
+            return LocalDateTime.parse(date, INPUT_DATE_FORMAT);
+        } catch (DateTimeParseException e) {
+            throw new IllegalArgumentException("Fecha inválida. Formato esperado: dd/MM/yyyy HHmm");
+        }
+    }
```
Maybe this method could be under parser instead for more clarity? 

Will there be more date formats that could be made available for future versions?

Frozennfishh commented on src/main/java/elchino/ui/Ui.java:
```diff
@@ -0,0 +1,59 @@
+     * Prints a message.
+     * @param message The message to print.
+     */
+    public void printMessage(String message) {
+        System.out.println(message);
+    }
```
Is this necessary?

limshaoqi replied:
yea

Frozennfishh commented on src/main/java/elchino/tasks/Task.java:
```diff
@@ -0,0 +1,81 @@
+ * Represents an abstract task that can be done.
+ */
+public abstract class Task {
+    protected final String description;
+    protected boolean isDone;
```
Maybe Class type could be included as a var of parent class Task for all tasks for more flexibility in the future?

Frozennfishh commented on src/main/java/elchino/tasks/Task.java:
```diff
@@ -0,0 +1,81 @@
+     * Retrieves the task type identifier.
+     * @return The task type identifier.
+     */
+    public abstract String getTaskType();
```
Referring to above comment, then this can read data from the class instead of printing manually, and may be easier to manage if down the line there may be methods that require reading class type (i.e. sort by task type)

Or a use of instanceOf to keep track of class data?

Frozennfishh commented on src/main/java/elchino/parser/Parser.java:
```diff
@@ -0,0 +1,33 @@
+package elchino.parser;
+
+import elchino.exceptions.*;
+import elchino.commands.*;
```
Perhaps be more consistent across all files where the line breaks are when displaying packages and imports

Frozennfishh commented on src/main/java/elchino/storage/Storage.java:
```diff
@@ -0,0 +1,128 @@
+        if (details.length < 3) {
+            System.out.println("⚠️ Error: Línea malformada en el archivo: " + line);
+            return null;
+        }
+
+        String type = details[0];
+        boolean isDone = details[1].equals("1");
+        String description = details[2];
+
+        try {
+            switch (type) {
+                case "T":
+                    Task todo = new Todo(description);
+                    if (isDone) todo.setDone();
+                    return todo;
+                case "D":
+                    if (details.length < 4) {
+                        System.out.println("Error: Línea malformada en el archivo: " + line);
+                        return null;
+                    }
+                    Task deadline = new Deadline(description, details[3]);
+                    if (isDone) deadline.setDone();
+                    return deadline;
+                case "E":
+                    if (details.length < 5) {
+                        System.out.println("Error: Línea malformada en el archivo: " + line);
+                        return null;
+                    }
+                    Task event = new Event(description, details[3], details[4]);
+                    if (isDone) event.setDone();
+                    return event;
+                default:
+                    System.out.println("Error: Tipo de tarea desconocido en el archivo: " + line);
+                    return null;
+            }
+        } catch (Exception e) {
+            System.out.println("Error: No se pudo procesar la tarea en el archivo: " + line);
+            return null;
+        }
+    }
```
Should this be under Parser as well?

Frozennfishh commented on src/main/java/elchino/tasks/TaskList.java:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * Prints all tasks in the list.
+     */
+    public void printTasks() {
+        if (tasks.isEmpty()) {
+            System.out.println("No hay tareas.");
+        } else {
+            System.out.println("Estas son tus tareas:");
+            for (int i = 0; i < tasks.size(); i++) {
+                System.out.println((i + 1) + ". " + tasks.get(i));
+            }
+        }
+    }
```
Maybe this can be under UI, but based on preference

Frozennfishh commented on src/main/java/elchino/tasks/Task.java:
```diff
@@ -0,0 +1,81 @@
+     * Converts the task into a string format for storage.
+     * @return The string format for storage.
+     */
+    public abstract String storeTask();
+
+    @Override
+    public String toString() {
+        return String.format("[%s] %s", isDone ? "X" : " ", description);
+    }
```
The format of these 2 methods are very similar, maybe can streamline into 1 to avoid confusion when reading and writing into files vs for printing?

Frozennfishh commented on src/test/java/elchino/tasks/TaskListTest.java:
```diff
@@ -0,0 +1,28 @@
+        Todo task2 = new Todo("Task 2");
+
+        // Test adding 1 task
+        taskList.addTask(task1);
+        assertEquals(1, taskList.getSize(), "TaskList should have 1 task after adding 1 task");
+
+        // Test adding 2 tasks
+        taskList.addTask(task2);
+        assertEquals(2, taskList.getSize(), "TaskList should have 2 tasks after adding 2 tasks");
+
+        // Verify that first task is task1
+        assertEquals("Task 1", taskList.getTask(1).getDescription(), "First task should match task1");
+
+        // Verify that second task is task2
+        assertEquals("Task 2", taskList.getTask(2).getDescription(), "Second task should match task2");
```
Can be more rigorous in testing, maybe adding more edge cases (mismatches, null lists etc.)

limshaoqi replied:
laze

Frozennfishh commented on src/test/java/elchino/parser/ParserTest.java:
```diff
@@ -0,0 +1,14 @@
+
+public class ParserTest {
+    @Test
+    public void testParseEmptyInput() throws ElchinoException {
+        Command command = Parser.parse("");
+        assertInstanceOf(InvalidCommand.class, command, "Empty input should return InvalidCommand");
+    }
```
Perhaps parser test should also test invalid inputs/different input cases

Frozennfishh reviewed (commented):
LGTM!
Spanish isn't my first language so it threw me off a bit at first, but structure of the code is good with no coding standard violations!
A few nits on import ordering and potential streamlining of certain methods being in certain classes but overall very coherent 👍

Commit:
Add a GUI to the chatbot

Commit:
Merge branch 'branch-Level-10'

Commit:
Parser does not check for missing arguments in commands.

This may cause unexpected errors when parsing user input, leading to bugs.

Let's add an 'assert' statement to ensure that commands requiring an argument do not proceed without one. Additionally, extract the 'requiresArgument' method to improve encapsulation and maintainability.

This ensures that the Parser fails fast when a required argument is missing, making debugging easier.

Commit:
Refactor Storage.java to use constants for error messages

limshaoqi committed:
Merge pull request #2 from limshaoqi/branch-A-Assertions

Parser does not check for missing arguments in commands.

Commit:
Merge branch 'master' of https://github.com/limshaoqi/ip

limshaoqi committed:
Merge pull request #3 from limshaoqi/branch-A-CodeQuality

Refactor Storage.java to use constants for error messages

Commit:
Merge branch 'master' of https://github.com/limshaoqi/ip

Commit:
Add HelpCommand

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Add UserGuide

Commit:
Add Ui.png

Commit:
Update CodeQuality

limshaoqi committed:
Delete src/main/java/elchino/data directory

Commit:
Update build.gradle

Commit:
Update UserGuide

Commit:
Merge branch 'master' of https://github.com/limshaoqi/ip
